url3426a97ee4_rtreffer_minidns_tgz-pJ8-de_measite_minidns_minidnsrepl_MiniDnsReplJ8/src/de/measite/minidns/recursive/RecursiveDNSClient.java

```diff
@@ -0,0 +1,597 @@
+        rootServerInetAddress('h', new int[]{0x2001, 0x0500, 0x0001, 0x0000, 0x0000, 0x000, 0x0000, 0x0053}),
+        rootServerInetAddress('i', new int[]{0x2001, 0x07fe, 0x0000, 0x0000, 0x0000, 0x000, 0x0000, 0x0053}),
+        rootServerInetAddress('j', new int[]{0x2001, 0x0503, 0x0c27, 0x0000, 0x0000, 0x000, 0x0002, 0x0030}),
+        rootServerInetAddress('l', new int[]{0x2001, 0x0500, 0x0003, 0x0000, 0x0000, 0x000, 0x0000, 0x0042}),
+        rootServerInetAddress('m', new int[]{0x2001, 0x0dc3, 0x0000, 0x0000, 0x0000, 0x000, 0x0000, 0x0035}),
+    };
+
+    int maxSteps = 128;
+
+    /**
+     * Create a new recursive DNS client using the global default cache.
+     */
+    public RecursiveDNSClient() {
+        super();
+    }
+
+    /**
+     * Create a new recursive DNS client with the given DNS cache.
+     *
+     * @param cache The backend DNS cache.
+     */
+    public RecursiveDNSClient(DNSCache cache) {
+        super(cache);
+    }
+
+    /**
+     * Recursively query the DNS system for one entry.
+     *
+     * @param queryBuilder The query DNS message builder.
+     * @return The response (or null on timeout/error).
+     * @throws IOException if an IO error occurs.
+     */
+    @Override
+    protected DNSMessage query(DNSMessage.Builder queryBuilder) throws IOException {
+        DNSMessage q = queryBuilder.build();
+        RecursionState recursionState = new RecursionState(this);
+        DNSMessage message = queryRecursive(recursionState, q);
+        return message;
+    }
+
+    private InetAddress getRandomIpv4RootServer() {
+        return IPV4_ROOT_SERVERS[insecureRandom.nextInt(IPV4_ROOT_SERVERS.length)];
+    }
+
+    private InetAddress getRandomIpv6RootServer() {
+        return IPV6_ROOT_SERVERS[insecureRandom.nextInt(IPV6_ROOT_SERVERS.length)];
+    }
+
+    private static InetAddress[] getTargets(Collection<? extends InternetAddressRR> primaryTargets,
+            Collection<? extends InternetAddressRR> secondaryTargets) {
+        InetAddress[] res = new InetAddress[2];
+
+        for (InternetAddressRR arr : primaryTargets) {
+            if (res[0] == null) {
+                res[0] = arr.getInetAddress();
+                // If secondaryTargets is empty, then try to get the second target out of the set of primaryTargets.
+                if (secondaryTargets.isEmpty()) {
+                    continue;
+                }
+            }
+            if (res[1] == null) {
+                res[1] = arr.getInetAddress();
+            }
+            break;
+        }
+
+        for (InternetAddressRR arr : secondaryTargets) {
+            if (res[0] == null) {
+                res[0] = arr.getInetAddress();
+                continue;
+            }
+            if (res[1] == null) {
+                res[1] = arr.getInetAddress();
+            }
+            break;
+        }
+
+        return res;
+    }
+
+    private DNSMessage queryRecursive(RecursionState recursionState, DNSMessage q) throws IOException {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This method contains 57 lines of code, not including blank lines or lines with only comments, Java punctuation characters, identifiers, or literals. By comparison, 99% of the methods in the CodeGuru reference dataset contain fewer lines of code. Large methods might be difficult to read and have logic that is hard to understand and test. We recommend that you simplify this method or break it into multiple methods.